### PR TITLE
[ENH] Implement FastBIC criterion to unsupervised

### PR DESCRIPTION
**Pull request by jshinm**
Fixes #19

Changes proposed in this pull request:

- Implement unit test for TwoMeans and FastBIC using scikit-learn Iris dataset

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: sktree/tree/tests/test_tree.py [hunk 1]
```diff
@@ -44,3 +60,20 @@ def test_unsupervisedtree():
 
     # a single decision tree does not fit well, but should still have a positive score
     assert score > 0.05
+
+def test_iris():
+    # Check consistency on dataset iris.
+    for (name, Tree), criterion in product(CLF_TREES.items(), CLF_CRITERIONS):
```
Comment (adam2392) [inline sktree/tree/tests/test_tree.py]: Use two instances of `pytest.mark.parametrize` instead will make it so that each of these is inherently a separate unit-test.

i.e.

```Python
@pytest.mark.parametrize('TREE', ...)
@pytest.mark.parametrize('CRITERION', ...)
def test_iris(TREE, CRITERION):
...
```

Comment (jshinm) [reply sktree/tree/tests/test_tree.py]: Thank you for the example! I saw this edit after I made the change. Does criterion need to be parameterized too as it's just string?

**Comment by jshinm** [comment]
@adam2392 Is `test_tree.py` update with Iris dataset enough for the parent PR? 

image

**Comment by adam2392** [comment]
Add it to the test forest too

**Comment by jshinm** [comment]
@adam2392 It seems like the forest passes, but here again with a lower accuracy

Screenshot from 2023-02-09 13-12-47

After adjusting raising accuracy threshold

Screenshot from 2023-02-09 13-18-27

File: sktree/tests/test_forest.py [hunk 2]
```diff
@@ -4,9 +4,24 @@
 from sklearn.datasets import make_blobs
 from sklearn.metrics import adjusted_rand_score
 from sklearn.utils.estimator_checks import parametrize_with_checks
+from sklearn import datasets
+from sklearn.metrics import accuracy_score
 
 from sktree import UnsupervisedRandomForest
 
+CLF_CRITERIONS = ("twomeans", "fastbic")
```
Comment (adam2392) [inline sktree/tests/test_forest.py:11]: Not classifiers.

File: sktree/tests/test_forest.py [hunk 3]
```diff
@@ -4,9 +4,24 @@
...
+
+FOREST_CLASSIFIERS = {
```
Comment (adam2392) [inline sktree/tests/test_forest.py]: Not classifiers.

File: sktree/tests/test_forest.py [hunk 4]
```diff
@@ -41,3 +56,28 @@ def test_urf():
...
+    clf.fit(iris.data, iris.target)
+    score = accuracy_score(clf.predict(iris.data), iris.target)
+    # score = clf.score(iris.data, iris.target)
+    assert score > 0.3, "Failed with criterion %s and score = %f" % (criterion, score)
+
+    clf = ForestClassifier(
+        n_estimators=10, criterion=criterion, max_features=2, random_state=1
+    )
+    clf.fit(iris.data, iris.target)
+    score = accuracy_score(clf.predict(iris.data), iris.target)
```
Comment (adam2392) [inline sktree/tests/test_forest.py]: What's the reasoning for using `accuracy_score`? Isn't it a classification measure? 

https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html

File: sktree/tree/tests/test_tree.py [hunk 5]
```diff
@@ -44,3 +60,21 @@ def test_unsupervisedtree():
...
+    # Check consistency on dataset iris.
+    for criterion in CLF_CRITERIONS:
+        clf = Tree(criterion=criterion, random_state=0)
+        clf.fit(iris.data, iris.target)
+        score = accuracy_score(clf.predict(iris.data), iris.target)
+        assert score > 0.3, "Failed with {0}, criterion = {1} and score = {2}".format(
+            name, criterion, score
+        )
+
+        clf = Tree(criterion=criterion, max_features=2, random_state=0)
```
Comment (adam2392) [inline sktree/tree/tests/test_tree.py]: What is the motivation for setting `max_features=2`?

**Review by adam2392 — CHANGES_REQUESTED** [review]
Left some comments. The lower accuracy is fine. It seems to be the case in URF. In UobliqueRF, it should perform better.

File: sktree/tree/tests/test_tree.py [hunk 6]
```diff
@@ -4,9 +4,25 @@
...
+# also load the iris dataset
+# and randomly permute it
+iris = datasets.load_iris()
+rng = np.random.RandomState(1)
+perm = rng.permutation(iris.target.size)
+iris.data = iris.data[perm]
+iris.target = iris.target[perm]
 
 @parametrize_with_checks([UnsupervisedDecisionTree(random_state=12)])
 def test_sklearn_compatible_estimator(estimator, check):
```
Comment (adam2392) [inline sktree/tree/tests/test_tree.py]: Please add a parametrization to check `UnsupervisedDecisionTree(..., criterion='fastbic')` and explicitly make the existing one `UnsupervisedDecisionTree(..., criterion='twomeans')` here too. The default currently is "twomeans", but we might change that to "fastbic".

Comment (adam2392) [reply sktree/tree/tests/test_tree.py]: See: https://scikit-learn.org/stable/modules/generated/sklearn.utils.estimator_checks.check_estimator.html

File: sktree/tests/test_forest.py (hunk 2 above)
Comment (jshinm) [reply sktree/tests/test_forest.py:11]: Thank you! Changed it to clusters

File: sktree/tests/test_forest.py (hunk 3 above)
Comment (jshinm) [reply sktree/tests/test_forest.py]: Same here

File: sktree/tests/test_forest.py (hunk 4 above)
Comment (jshinm) [reply sktree/tests/test_forest.py]: Thank you for this! I followed your `adjusted_rand_score method`

File: sktree/tree/tests/test_tree.py (hunk 5 above)
Comment (jshinm) [reply sktree/tree/tests/test_tree.py]: I dropped this arg

File: sktree/tree/tests/test_tree.py (hunk 6 above)
Comment (jshinm) [reply sktree/tree/tests/test_tree.py]: Added!

**Comment by jshinm** [comment]
@adam2392 adjusted_rand_score on fastbic yields negative number. I set it on assert so that the code builds and tests, but is this also expected or is there something wrong with fastbic implementation?

**Comment by jshinm** [comment]
@adam2392 setting impurity to negative yielded the same scores.

**Comment by adam2392** [comment]
Maybe merge this PR into your other branch and I will take a look. I can't checkout your code in this PR because it is in your personal fork.

It's easier to see everything in one PR usually, unless the code is very massively changed.

**Comment by adam2392** [comment]
Lmk when you merge it
